leshan-core/src/test/java/org/eclipse/leshan/core/util/TimestampUtilTest.java

```diff
@@ -0,0 +1,72 @@
+package org.eclipse.leshan.core.util;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThrows;
+
+import java.math.BigDecimal;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TimestampUtilTest {
+
+    @Test
+    public void test_instant_to_big_decimal_conversion() {
+        List<Instant> inputList = Arrays.asList(null, Instant.EPOCH, Instant.MIN, Instant.MAX,
+                Instant.ofEpochSecond(1664832119, 123456789));
+
+        List<BigDecimal> expectedOutputList = Arrays.asList(null, BigDecimal.ZERO,
+                BigDecimal.valueOf(-31557014167219200L),
+                BigDecimal.valueOf(31556889864403199L).add(BigDecimal.valueOf(0.999999999)),
+                BigDecimal.valueOf(1664832119L).add(BigDecimal.valueOf(0.123456789)));
+
+        List<BigDecimal> outputList = inputList.stream().map(TimestampUtil::fromInstant).collect(Collectors.toList());
+
+        Assert.assertEquals(expectedOutputList, outputList);
+    }
+
+    @Test
+    public void test_big_decimal_to_instant_conversion() {
+        List<BigDecimal> inputList = Arrays.asList(null, BigDecimal.ZERO, BigDecimal.valueOf(-31557014167219200L),
+                BigDecimal.valueOf(31556889864403199L).add(BigDecimal.valueOf(0.999999999)),
+                BigDecimal.valueOf(1664832119L).add(BigDecimal.valueOf(0.123456789)));
+
+        List<Instant> expectedOutputList = Arrays.asList(null, Instant.EPOCH, Instant.MIN, Instant.MAX,
+                Instant.ofEpochSecond(1664832119, 123456789));
+
+        List<Instant> outputList = inputList.stream().map(TimestampUtil::fromSeconds).collect(Collectors.toList());
+
+        Assert.assertEquals(expectedOutputList, outputList);
+    }
+
+    @Test
+    public void test_too_precise_big_decimal_to_instant_conversion() {
+        long seconds = 1664832119L;
+        long nanoseconds = 112233445L;
+        long picoseconds = 112233445566L;
+
+        BigDecimal tooPreciseValue = BigDecimal.valueOf(seconds)
+                .add(BigDecimal.valueOf(picoseconds).divide(BigDecimal.valueOf(1E12)));
+
+        Instant convertedValue = TimestampUtil.fromSeconds(tooPreciseValue);
+
+        Assert.assertEquals(seconds, convertedValue.getEpochSecond());
+        Assert.assertNotEquals(picoseconds, convertedValue.getNano());
+        Assert.assertEquals(nanoseconds, convertedValue.getNano());
+    }
```

Comment:
Should we raise an exception instead ?

Reply:
I understand your point of view.
But I also think this could also hide issue, and so generally I prefer to just raise exception by default to show there is something not expected and then let user decide, how to handle this : 
 - stop using too much precision at client side.
 - lost precision at server side.
 - or anything else...